### DISCUSSION TITLE
Configuring nvim-dap for C++ and Typescript in lang Extras

### DISCUSSION THREAD

**Discussion by ifonajs** [LazyVim/LazyVim#2548] [General]
Hi,

How can I override the default opts for [nvim-dap configured for Clangd lang Extras](https://www.lazyvim.org/extras/lang/clangd#nvim-dap-optional)?

I tried defining my own function assuming the default one would get overridden. Didn't work.

I checked [lazy.nvim documentation](https://github.com/folke/lazy.nvim/blob/main/doc/lazy.nvim.txt#L142C18-L142C54) and opts are specified to take `table or fun(LazyPlugin, opts:table)`, never function(). It seems to be the same for [Typescript lang Extras](https://www.lazyvim.org/extras/lang/typescript#nvim-dap-optional). Usually if opts are just a function(), then this function returns a table of opts but this is for some reason not the case here. Why is configuring `nvim-dap` exception to this?

**Comment by ifonajs** [comment]
Anyone has any idea? How do I override these opts:
```lua
opts = function()
  local dap = require("dap")
  if not dap.adapters["codelldb"] then
    require("dap").adapters["codelldb"] = {
      type = "server",
      host = "localhost",
      port = "${port}",
      executable = {
        command = "codelldb",
        args = {
          "--port",
          "${port}",
        },
      },
    }
  end
  for _, lang in ipairs({ "c", "cpp" }) do
    dap.configurations[lang] = {
      {
        type = "codelldb",
        request = "launch",
        name = "Launch file",
        program = function()
          return vim.fn.input("Path to executable: ", vim.fn.getcwd() .. "/", "file")
        end,
        cwd = "${workspaceFolder}",
      },
      {
        type = "codelldb",
        request = "attach",
        name = "Attach to process",
        processId = require("dap.utils").pick_process,
        cwd = "${workspaceFolder}",
      },
    }
  end
end
```

**Comment by dpetka2001** [comment]
Something like the following should work in theory
```lua
return {
  {
    "mfussenegger/nvim-dap",
    opts = function()
      local dap = require("dap")
      dap.adapters.codelldb.port = 13000
      dap.adapters.codelldb.type = "testing"
    end,
  },
}
```
If you then do `:lua vim.print(require("dap").adapters["codelldb"])` you should be able to verify the modified settings. You can also run this command before you make the changes to see what the actual configuration is (it should be the same as in the LazyVim docs). 

Same can be done about changing values in `dap.configurations.c[1].name = "test"` for the first table (which has `name = "Launch file",`) and `dap.configurations.c[2].name = "whatever"` for the second table (which has `name = "Attach to process",`).

**Comment by ifonajs** [comment]
Didn't work as many other things. It keeps ignoring whatever I do with opts. I will open an issue.

**Reply by dpetka2001** [reply]
What didn't work exactly? I tried it and it changed the options although I don't do C/C++. I just checked the configuration of `nvim-dap` that it had changed. What options exactly do you want to change and to what?

Another alternative would be to just copy the [clangd](https://github.com/LazyVim/LazyVim/blob/main/lua/lazyvim/plugins/extras/lang/clangd.lua) extra to your plugins folder and don't enable it via `LazyExtras`, but instead incorporate directly the changes that you want.

No need to open an issue as this is a user configuration issue and when you create an issue, the template for bugs clearly states that questions of "How do I do something" belong in the Discussions section.

**Comment by ifonajs** [comment]
Found the root cause of my problems. I was installing extras with a file defined in ~/.config/nvim/lua/plugins/extras.lua. When I moved importing of extras modules to ~/.config/nvim/lua/config/lazy.lua, now I'm able to change configuration of nvim-dap. Seems like importing extras plugins overridden whatever I defined earlier except when I was simply changing opts. Then it did attempt merging.

Thanks for the help.

**Reply by dpetka2001** [reply]
Well yes, you can't import the extras from a file in `/lua/plugins/` directory. It even mentions it in the [docs](https://www.lazyvim.org/extras/lang/clangd)
> You can enable the extra with the :LazyExtras command. Plugins marked as optional will only be configured if they are installed.
Alternatively, you can add it to your lazy.nvim imports
lua/config/lazy.lua
require("lazy").setup({
  spec = {
    { "LazyVim/LazyVim", import = "lazyvim.plugins" },
    { import = "lazyvim.plugins.extras.lang.clangd" },
    { import = "plugins" },
  },
})

If your problem is solved, please also close the issue you created.
